File: src/go/problems/my_sqrt_test.go
// https://leetcode.com/problems/sqrtx

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type MySqrt struct{}

func TestMySqrt(t *testing.T) {
	gen := core.TestSuite[MySqrt]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param(4).Result(2)
	}).Add(func(tc *core.TestCase) {
		tc.Param(8).Result(2)
	}).Add(func(tc *core.TestCase) {
		tc.Param(13284790).Result(3644)
	}).Run(t)
}

func (MySqrt) Solution(x int) int {
	left := 0
	right := x
	for left <= right {
		root := (left + right) / 2
		target := root * root

		if target == x {
			return root
		} else if target < x {
			left = root + 1
		} else {
			right = root - 1
		}
	}

	return right
}

File: src/challenges/python/problems/my_sqrt_test.py
# https://leetcode.com/problems/sqrtx

from core.problem_base import *

class MySqrt(ProblemBase):
    def Solution(self, x: int) -> int:
        left = 0
        right = x
        while left <= right:
            root = (left + right) // 2
            target = root * root

            if target == x:
                return root
            elif target < x:
                left = root + 1
            else:
                right = root - 1

        return right

if __name__ == '__main__':
    TestGen(MySqrt) \
        .Add(lambda tc: tc.Param(4).Result(2)) \
        .Add(lambda tc: tc.Param(8).Result(2)) \
        .Add(lambda tc: tc.Param(13284790).Result(3644)) \
        .Add(lambda tc: tc.Param(0).Result(0)) \
        .Add(lambda tc: tc.Param(1).Result(1)) \
        .Add(lambda tc: tc.Param(2).Result(1)) \
        .Run()
